src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -766,110 +762,138 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
-                        // do the flush if our decrement caused the pending count to reach 0: if it's still higher than zero, then there's
-                        // at least one other pending writer who can handle the flush.  Worst case, we pay for a flush that ends up being
-                        // a nop.  Note: we explicitly do not pass in the cancellationToken; if we're here, it's almost certainly because
-                        // cancellation was requested, and it's because of that cancellation that we need to flush.
-                        LogExceptions(FlushAsync(cancellationToken: default));
-                    }
+                WriteBytes = writeBytes;
 
-                    throw;
-                }
-                Interlocked.Decrement(ref _pendingWriters);
+                _cancellationToken = cancellationToken;
+                _cancellationRegistration = cancellationToken.UnsafeRegister(s => ((WriteQueueEntry)s!).OnCancellation(), this);
             }
 
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (_abortException != null)
+            public int WriteBytes { get; }
+
+            private void OnCancellation()
             {
-                _writerLock.Exit();
-                ThrowRequestAborted(_abortException);
+                SetCanceled(_cancellationToken);
             }
 
-            // Flush waiting state, then invoke the supplied action.
-            try
+            public bool TryDisableCancellation()
             {
-                // If there is a pending write that was canceled while in progress, wait for it to complete.
-                if (_inProgressWrite != null)
-                {
-                    await new ValueTask(_inProgressWrite).ConfigureAwait(false); // await ValueTask to minimize number of awaiter fields
-                    _inProgressWrite = null;
-                }
+                _cancellationRegistration.Dispose();
+                return !Task.IsCanceled;
+            }
 
-                // If the buffer has already grown to 32k, does not have room for the next request,
-                // and is non-empty, flush the current contents to the wire.
-                int totalBufferLength = _outgoingBuffer.Capacity;
-                if (totalBufferLength >= UnflushedOutgoingBufferSize)
-                {
-                    int activeBufferLength = _outgoingBuffer.ActiveLength;
-                    if (writeBytes >= totalBufferLength - activeBufferLength && activeBufferLength > 0)
-                    {
-                        // We explicitly do not pass cancellationToken here, as this flush impacts more than just this operation.
-                        await new ValueTask(FlushOutgoingBytesAsync()).ConfigureAwait(false); // await ValueTask to minimize number of awaiter fields
-                    }
-                }
+            public abstract bool InvokeWriteAction(Memory<byte> writeBuffer);
```

Comment:
Nit: `protected` ?

Reply:
It's invoked outside the class, see line 865